src/python/pants/core/goals/init.py

```diff
@@ -0,0 +1,224 @@
+from pants.engine.fs import CreateDigest, Digest, DigestContents, FileContent, PathGlobs, Workspace
+from pants.engine.goal import Goal, GoalSubsystem, LineOriented
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import collect_rules, goal_rule, rule
+from pants.engine.target import Targets
+from pants.engine.unions import UnionMembership, union
+from pants.util.frozendict import FrozenDict
+from pants.util.meta import frozen_after_init
+
+logger = logging.getLogger(__name__)
+
+
+@union
+class PutativeTargetsRequest(metaclass=ABCMeta):
+    pass
+
+
+@frozen_after_init
+@dataclass(order=True, unsafe_hash=True)
+class PutativeTarget:
+    """A potential target to add, detected by various heuristics."""
+
+    # Note that field order is such that the dataclass order will be by address (path+name).
+    path: str
+    name: str
+    type_alias: str
+    # The sources that triggered creating of this putative target.
+    sources: Tuple[str, ...]
+    # Note that we generate the BUILD file target entry exclusively from these kwargs (plus the
+    # type_alias), not from the name and sources fields above, which are broken out for other uses.
+    # This allows the creator of instances of this class to control whether the generated
+    # target should assume default kwarg values or provide them explicitly.
+    kwargs: FrozenDict[str, Union[str, int, bool, Tuple[str, ...]]]
+    build_file_name: str
+
+    def __init__(
+        self,
+        path: str,
+        name: str,
+        type_alias: str,
+        sources: Iterable[str],
+        *,
+        kwargs: Union[None, Mapping[str, Union[str, int, bool, Tuple[str, ...]]]] = None,
+        build_file_name="BUILD",
+    ) -> None:
+        self.path = path
+        self.name = name
+        self.type_alias = type_alias
+        self.sources = tuple(sources)
+        self.kwargs = FrozenDict(kwargs or {})
+        self.build_file_name = build_file_name
+
+    @property
+    def build_file_path(self) -> str:
+        return os.path.join(self.path, self.build_file_name)
+
+    @property
+    def address_spec(self) -> str:
+        return f"{self.path}:{self.name}"
+
+    def generate_build_file_stanza(self) -> str:
+        indent = "  "
+
+        if self.kwargs:
+            kwargs_str_parts = [f"\n{indent}{k}={repr(v)}" for k, v in self.kwargs.items()]
+            kwargs_str = ",".join(kwargs_str_parts) + ",\n"
+        else:
+            kwargs_str = ""
+        return f"{self.type_alias}({kwargs_str})\n"
+
+
+class PutativeTargets(DeduplicatedCollection[PutativeTarget]):
+    sort_input = True
+
+    @classmethod
+    def merge(cls, tgts_iters: Iterable["PutativeTargets"]) -> "PutativeTargets":
+        all_tgts: List[PutativeTarget] = []
+        for tgts in tgts_iters:
+            all_tgts.extend(tgts)
+        return cls(all_tgts)
+
+
+class InitSubsystem(LineOriented, GoalSubsystem):
+    name = "init"
+    help = "Generate config for working with Pants."
```

Comment:
We should likely change the name to make it clear both to implementers and users that this goal is intended to be able to be run multiple times, rather than only once to "initialize" the repo.

It also might be good to have a "preview" mode that reports what would have happened (possibly with a particular failing exit code if changes would have been made). This smells a bit like the `lint`/`fmt` split: do we want to be able to `lint` for changes that might be made here (even if actually commiting the changes happens via a separate goal)?

Reply:
Good idea, will do.

Reply:
Chosen name is "tailor".